design/plugin-versioning.md

```diff
@@ -0,0 +1,201 @@
+The Server will also need to be adapted to recognize this new plugin Kind and to serve the new implementation.
+
+### Plugin Manager
+
+The plugin manager is responsible for managing the lifecycle of plugins.
+It provides an interface which is used by Velero to retrieve an instance of a plugin kind with a specific name (e.g. ObjectStore with the name "velero.io/aws").
+The manager contains a registry of all available plugins which is populated during the main Velero server startup.
+When the plugin manager is requested to provide a particular plugin, it checks the registry for that plugin kind and name.
+If it is available in the registry, the manager retrieves a `RestartableProcess` for the plugin binary, creating it if it does not already exist.
+That `RestartableProcess` is then used by individual restartable implementations of a plugin kind (e.g. `restartableObjectStore`, `restartableVolumeSnapshotter`).
+
+As new plugin versions are added, the plugin manager will be modified to retrieve the latest version of a plugin kind, adapting earlier versions of the plugin kind if possible.
+This is to allow the remainder of the Velero codebase to assume that it will always interact with the latest version of a plugin.
+The manager will check the registry for the plugin kind and name, and if the requested version is not found, it will attempt to fetch previous versions of the plugin kind and adapt it if possible.
+It will be up to the author of new plugin versions to determine whether a previous version of a plugin can be adapted to work with the interface of the new version.
+
+```
+// GetObjectStore returns a restartableObjectStore for name.
+func (m *manager) GetObjectStore(name string) (v2.ObjectStore, error) {
+    name = sanitizeName(name)
+
+    restartableProcess, err := m.getRestartableProcess(framework.PluginKindObjectStoreV2, name)
+    if err != nil {
+        // Check if plugin was not found
+        if errors.Is(err, pluginNotFoundError) {
+            // Try again but with previous version
+            restartableProcess, err := m.getRestartableProcess(framework.PluginKindObjectStore, name)
+            if err != nil {
+                // No v1 version found, return
+                return nil, err
+            }
+
+            r := newRestartableObjectStore(name, restartableProcess)
+            // Adapt v1 plugin to v2
+            return newAdaptedV1ObjectStore(r), nil
+        } else {
+            return nil, err
+        }
+    }
+    return newRestartableObjectStoreV2(name, restartableProcess), nil
+}
+```
+
+If the previous version is not available, or can not be adapted to the latest version, an error will be returned as is currently the case when a plugin implementation for a particular kind and provider can not be found.
+
```

Comment:
Once we have more than 2 versions (i.e. we're on v3, v4, etc.) the above might become somewhat unwieldy with the nesting of conditionals. It might make sense to refactor this into a method that takes a slice of versions/apis/etc. to check. At each entry, behavior would either be to return the found plugin, go to the next iteration, or return an error, depending on whether the current call returned without error, returned a `pluginNotFoundError`, or returned some other error.

Reply:
@phuongatemc Yep, it is very verbose. Unfortunately the plugins require a lot of code to be written and every new version is essentially a new plugin kind :( I think this would be a great thing to improve and refactor in future but I don't think we can include it in the scope of this work.

@sseago - I think the changes to the code block above fit the idea you suggested where we iterate over a slice of plugin kinds (sorted by version in descending order). Let me know if you have other ideas or suggestions though :+1: